aiosmtpd/smtp.py

```diff
@@ -352,28 +352,33 @@ def _set_rset_state(self):
         if not hostname:
             await self.push('501 Syntax: EHLO hostname')
             return
+
+        response = []
         self._set_rset_state()
         self.session.extended_smtp = True
-        await self.push('250-%s' % self.hostname)
+        response.append('250-%s' % self.hostname)
         if self.data_size_limit:
-            await self.push('250-SIZE %s' % self.data_size_limit)
+            response.append('250-SIZE %s' % self.data_size_limit)
             self.command_size_limits['MAIL'] += 26
         if not self._decode_data:
-            await self.push('250-8BITMIME')
+            response.append('250-8BITMIME')
         if self.enable_SMTPUTF8:
-            await self.push('250-SMTPUTF8')
+            response.append('250-SMTPUTF8')
             self.command_size_limits['MAIL'] += 10
         if self.tls_context and not self._tls_protocol:
-            await self.push('250-STARTTLS')
+            response.append('250-STARTTLS')
         if hasattr(self, 'ehlo_hook'):
             warn('Use handler.handle_EHLO() instead of .ehlo_hook()',
                  DeprecationWarning)
             await self.ehlo_hook()
         status = await self._call_handler_hook('EHLO', hostname)
         if status is MISSING:
             self.session.host_name = hostname
-            status = '250 HELP'
-        await self.push(status)
+            response.append('250 HELP')
```

Comment:
For loop has to be outside the if block. It is now executed only when the hook is missing.

Reply:
IIUC that's the point - if there's an EHLO hook, there should be a response - otherwise we should assume that the EHLO hook is going to specify all of the above. @cnicodeme is that correct?

Reply:
Okay, yeah, refreshing myself with the issue at hand, yeah, that's the case, and this is definitely the right place for it.

Reply:
yeah. got now. However not clear on the responsibilities. Whether the hook is supposed to push the status and other messages too. If so it cannot return any status.

Reply:
The hook shouldn't be handling any communication, it should only be returning status lines that will then be passed to the client.

Reply:
Accidentally deleted my earlier comment, ugh -_-

Retyping from memory: If we expect the hook to be "sane"-ly constructed (see #154 ) then the "if" there should be `if status is MISSING or status is None:`

If the hook wants to handle the communication -- after all, we passed `self` to it -- then it should return "something" (other than `MISSING` or `None`). In which case, the `else:` clause should not be there.

Reply:
Resolved by redefining a completely new behavior for `handle_EHLO`.

This new behavior is implemented in a way that's backward-compatible, and it has been documented.